### PR TITLE
Enable noImplicitThis

### PR DESCRIPTION
**Pull request by mohsen1**
**What kind of change does this PR introduce?**
Types: Trying to enable `noImplicitThis` tsconfig

**Did you add tests for your changes?**
no

**Does this PR introduce a breaking change?**

**What needs to be documented once your changes are merged?**

### CODE REVIEW

**Comment by sandersn** [comment]
Depends on Microsoft/Typescript#13308. Otherwise, the only way I can think of to specify this types is to use a type annotation on a function expression:

```js
/** @type {(this: number) => void} */
function f() {
    this.toFixed()
}
// OR //
/** @type {function(this:number)} */
function f() {
    this.toFixed()
}
```

**Comment by mohsen1** [comment]
I think webpack doesn't want to use TS syntax for types. I can wait for `@this` support lands in TypeScript. Do you know when that can happen?

**Comment by mohsen1** [comment]
I was thinking if we can make util.deprecate a generic so class type follows in... I need to play with that idea

**Comment by sandersn** [comment]
@mohsen1 Microsoft/TypeScript#24927 will support `@this` when it's merged.

**Comment by sandersn** [comment]
Microsoft/TypeScript#24927 is in master, so typescript@next will have `@this` starting tomorrow.

**Comment by mohsen1** [comment]
@sandersn awesome work! most error are gone! There are these two weird errors that I couldn't go around:


```
lib/Module.js:365:22 - error TS2345: Argument of type 'SortableSet' is not assignable to parameter of type 'ArrayLike<any>'.
  Property 'length' is missing in type 'SortableSet'.

365             return Array.from(this._chunks, fn);
                                  ~~~~~~~~~~~~
```
`SortableSet` extends `Set` which is utterable. why `Array.from` is not happy with `this._chunks`?



```
lib/webpack.js:72:40 - error TS2683: 'this' implicitly has type 'any' because it does not have a type annotation.

72 webpack.validate = validateSchema.bind(this, webpackOptionsSchema);
                                          ~~~~
```

Global `this`! Did you guys cover this under `@this` support?

**Comment by sandersn** [comment]
1. I can't repro this error from the editor. When I goto-def on Array.from there, it goes to lib.es2015.iterable.d.ts. It looks like CI is only seeing definitions from lib.es2015.core.d.ts. However, the tsconfig specifies "lib": ["es2017", "dom"]. Is that the tsconfig that CI is using too? Does it help to manually specify "es2015.iterable" in that list?
2. `@this` behaves exactly like `this` parameters in typescript, so it can't help you directly with accesses to global `this`. This is by design, for better or worse. We thought that global this should require extra care once the flag is turned on. Specifically, casts will not suffice. Here's how we worked around our global this access in Typescript: `(/** @this {any} */ function () { return this })()` (in Microsoft/TypeScript#11265).

Previously, in Microsoft/TypeScript#9580, I used `new Function("return this")()`. Unfortunately this method breaks execution under Atom (and other restricted environments with Content Security Policy set to disallow `eval` and friends). This may not apply to webpack though.

File: lib/util/SortableSet.js [hunk 1]
```diff
@@ -3,6 +3,7 @@
 /**
  * A subset of Set that offers sorting functionality
  * @template T item type in set
+ * @extends {Set<T>}
```
Comment (Legends) [inline lib/util/SortableSet.js]: [Typings Guide -section `Class`](https://github.com/webpack/webpack/issues/7122)
Or is this rule applicable only if you inherit from custom types?

Comment (mohsen1) [reply lib/util/SortableSet.js]: I wish I didn't have to have `@extend` but generic type is not inferred

Comment (sandersn) [reply lib/util/SortableSet.js]: Typescript needs to know that SortableSet<T> extends Set<T>, and there's no way to specify the type variables in Javascript. Probably the rule needs an exception.

Comment (sandersn) [reply lib/util/SortableSet.js]: By the way, Microsoft/TypeScript#25101 will be fixed in typescript@next tomorrow.

**Comment by mohsen1** [comment]
@sandersn Thanks for handing that case. There are two new errors that popped up which I can't go around them:

# 1
```
lib/Chunk.js:177:48 - error TS2339: Property 'getRuntimeChunk' does not exist on type 'ChunkGroup | Entrypoint'.
  Property 'getRuntimeChunk' does not exist on type 'ChunkGroup'.

177  		return chunkGroup.isInitial() && chunkGroup.getRuntimeChunk() === this;
     		                                            ~~~~~~~~~~~~~~~
```
Per my types, `chunkGroup` is `ChunkGroup | EntryPoint`. The `isInitial` method exists in both classes. `EntryPoint#isInitial` returns `true` while `ChunkGroup#isInitial` returns `false`. Is there a way of using the type system to infer the type based on this function return type? In TypeScript we could make the return type `is EntryPoint` but it seems that's not supported in Salsa.

# 2
```
lib/Module.js:363:22 - error TS2345: Argument of type 'SortableSet<Chunk>' is not assignable to parameter of type 'ArrayLike<any>'.
  Property 'length' is missing in type 'SortableSet<Chunk>'.

363  		return Array.from(this._chunks, fn);
     		                  ~~~~~~~~~~~~
```

This is puzzling because `this._chunks` is a `SortableSet` which extends `Set` so it's an iterable. `Array.from` expects an `ArrayLike`. 

`Array.from` does work with iterables:
```js
function* iter() {
  for (const i of [1,2,3]) yield I
}
Array.from(iter(), i => i * 2)
// => [2, 4, 6]

```

**Comment by sandersn** [comment]
A couple of comments from the bus:
1. I’ll take a look. Probably the type guard code checks the return type nose directly and forgets to check `@returns`. 
2. I couldn’t repro this in my editor. Can you? If so, how many overloads does Array.from report having? I think there should be 3 if es2105.iterable is correctly included.

**Comment by sandersn** [comment]
Filed Microsoft/TypeScript#25127 to track the type guard bug.

**Comment by sandersn** [comment]
The bogus non-iterable error is a Very Bad Error Message that sometimes happens with overloads. Basically, the *first* overload should have given the error, but the compiler tries all four and gives the error that you would get if the *last* overload were the only one.

Here's how I debugged this problem:

1. Copy all the overloads of Array.from from es2015.core and es2015.iterable into declarations.d.ts and rename them to a global `declare function af`.
2. Copy the offending call on the line above as `af(this._chunks, fn)` and check that the error is the same.
3. Go back to the copied declarations and manually eliminate the overloads that don't apply. In this case, the 1-parameter ones obviously don't, and the ArrayLike-only one doesn't. This gets us back down to one signature.
4. Check the error message on `af(this._chunks. fn)`. Now it says fn's type `function(any, any, Set<any>): void` isn't assignable to the mapper function `function(Chunk, number): void`.
5. That's because `Array.from` will never pass a third argument to `fn`. So I think the correct fix is to make the type of `fn: function(any, any): void`.

Note that you can actually make `fn: function(Chunk, number): void` and the return type `Array<void>` based on the types I saw inferred from Array.from.

**Comment by sandersn** [comment]
Type guards will be working in `typescript@next` tomorrow.

**Comment by mohsen1** [comment]
@sandersn we include es2017 lib which should have iterables. Can you try running tsc? It's still happening on my setup and on TravisCI

**Comment by sandersn** [comment]
@mohsen1 I got it to repro in my editor and discovered that it's a bad error message, not a problem with missing es2015.iterable at all. See my 2nd-to-last comment.

**Comment by mohsen1** [comment]
@sandersn Thanks for explaining. changing `fn` type resolved the issue. The bug with not narrowing types with `chunkGroup.isInitial()` is still happening

**Comment by sandersn** [comment]
@mohsen1 You still need to change `Entrypoint.isInitial` from `@returns {true}` to `@returns {this is Entrypoint}` on Entrypoint.js:30.

Note that I tried adding `@returns {this is ChunkGroup}` to `ChunkGroup.isInitial` and this made the type of `isInitial(): this is Entrypoint | ChunkGroup`, which takes us back to square one. I think this is a limitation of this-type-guards.

File: lib/Chunk.js [hunk 2]
```diff
@@ -183,7 +184,11 @@ class Chunk {
 	hasRuntime() {
 		for (const chunkGroup of this._groups) {
 			// We only need to check the first one
-			return chunkGroup.isInitial() && chunkGroup.getRuntimeChunk() === this;
+			return (
+				chunkGroup.isInitial() &&
+				// prettier-ignore
```
Comment (mohsen1) [inline lib/Chunk.js]: https://github.com/prettier/prettier/issues/4799

**Pull request closed without merging**